File: cousins_of_bt.java
//T.C- O(N)- in worst case we might have to visit all nodes 
//T.C-O(N)- space utilized by the queue
import java.util.ArrayList;
import java.util.LinkedList;

//   Definition for a binary tree node.
  public class cousins_of_bt {
      int val;
      cousins_of_bt left;
      cousins_of_bt right;
      cousins_of_bt() {}
      cousins_of_bt(int val) { this.val = val; }
      cousins_of_bt(int val, cousins_of_bt left, cousins_of_bt right) {
          this.val = val;
          this.left = left;
          this.right = right;
      }
  }
 

class Solution {
    public boolean isCousins(cousins_of_bt root, int x, int y) {
        if(root==null){
            return false;
        }
        LinkedList<cousins_of_bt> queue= new LinkedList<>();
        queue.add(root);
        while(!queue.isEmpty()){
            int size= queue.size();
            ArrayList<Integer>level= new ArrayList<>();
            for(int i=0; i< size;i++){
                cousins_of_bt temp=queue.remove();
                if(temp.left!=null && temp.right!=null){
                    if(temp.left.val==x && temp.right.val==y){
                        return false;
                    }
                    if(temp.right.val==x && temp.left.val==y){
                        return false;
                    }
                }
                level.add(temp.val);
                if(temp.left!=null){
                    queue.add(temp.left);
                }
                if(temp.right!=null){
                    queue.add(temp.right);
                }
            }
            if(level.contains(x) && level.contains(y)){
                return true;
            }
            
        }
        return false;
        
    }
}

File: Cousins-of-BT.py
# Time Complexity - O(N)
# Space Complexity - O(N)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque

class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        level_map = dict()
        queue = deque()
        queue.append([root, 1, None])

        while queue:
            value, level, parent = queue.popleft()
            if value.val == x or value.val == y:
                if level in level_map:
                    level_map[level].append([value.val, parent.val])
                else:
                    level_map[level] = [[value.val, parent.val if parent != None else None]]
            
            if value.left:
                queue.append([value.left, level+1, value])
            if value.right:
                queue.append([value.right, level+1, value])

        print(level_map)
        if len(level_map.keys()) > 1:
            return False
        else:
            level = level_map.keys()
            for key in level:
                key = int(key)
            if level_map[key][0][1] != level_map[key][1][1]:
                return True
            return False
        